feat(cgi): Display mount information in monitoring

rolysr opened the pull request:
This commit updates the CGI module to display mount point information for each mounted client in the monitoring interface. This enhancement improves visibility into cluster conditions in customer-deployed environments, aiding in identifying root causes of reported issues more effectively.

rolysr committed:
feat(cgi): Display mount information in monitoring

This commit updates the CGI module to display mount point information
for each mounted client in the monitoring interface. This enhancement
improves visibility into cluster conditions in customer-deployed
environments, aiding in identifying root causes of reported issues
more effectively.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

uristdwarf commented on src/master/matoclserv.cc:
```diff
@@ -1257,18 +1258,28 @@ void matoclserv_chunks_health(matoclserventry *eptr, const uint8_t *data, uint32
-	if (length!=0 && length!=1) {
+
+	auto isMountInfoRequired = [](bool includeMountInfo, const matoclserventry *eaptr) -> bool {
+		return includeMountInfo && eaptr->version >= saunafsVersion(4, 9, 0) && eaptr->sesdata &&
+		       eaptr->sesdata->mountinfo.c_str() && eaptr->sesdata->mountinfo.size() > 0;
```
Why `mountinfo.c_str()`? Why would `mountinfo.size() > 0` not be enough?

rolysr replied:
Done.

uristdwarf commented on src/master/matoclserv.cc:
```diff
@@ -1905,6 +1926,14 @@ void matoclserv_register_config(matoclserventry *eptr, const uint8_t *data,
 }
 
+void matoclserv_update_mount_info(matoclserventry *eptr, const uint8_t *data, uint32_t length) {
+	std::string moun_info;
```
Small typo

uristdwarf commented on src/cgi/sfs.cgi.in:
```diff
@@ -2311,7 +2311,7 @@ if "MS" in sectionset:
         s = socket.socket()
         s.connect((masterhost, masterport))
         if masterversion >= (1, 6, 26):
-            mysend(s, struct.pack(">LLB", CLTOMA_SESSION_LIST, 1, 1))
+            mysend(s, struct.pack(">LLBB", CLTOMA_SESSION_LIST, 2, 1, 1))
         else:
             mysend(s, struct.pack(">LL", CLTOMA_SESSION_LIST, 0))
```
Maybe there should be another check for masterversion here, for 4.9.0.
Also, all the checks should be greater than 5.0.0, because this PR won't be included in the release

uristdwarf reviewed (commented):
Initial review, I haven't yet tested this change

uristdwarf reviewed (dismissed):
I got a double free error and it core dumped:
![image](https://github.com/user-attachments/assets/de7a7620-d4b3-4df3-b16f-8c61c073a59d)

Solved. The issue was in packet size calculation when sending data from master server to cgi server.

ralcolea commented on src/master/matoclserv.cc:
```diff
@@ -1257,18 +1258,29 @@ void matoclserv_chunks_health(matoclserventry *eptr, const uint8_t *data, uint32
+		       eaptr->sesdata->mountinfo.size() > 0;
+	};
+
+	if (length!=0 && length!=1 && length!=2) {
```
Consider applying the current code style to the modified lines. For instance, add a whitespace to separate the operators and the operands.

rolysr replied:
Done.

ralcolea commented on src/master/matoclserv.cc:
```diff
@@ -1257,18 +1258,29 @@ void matoclserv_chunks_health(matoclserventry *eptr, const uint8_t *data, uint32
 	}
 	if (length==0) {
 		vmode = 0;
+	} else if (length==1) {
```
Same suggestion here.

rolysr replied:
Done.

ralcolea commented on src/mount/fuse/mount_config.cc:
```diff
@@ -119,6 +119,7 @@ struct fuse_opt gSfsOptsStage2[] = {
 };
 
 void initialize_opts_name_values() {
+	std::unique_lock<std::mutex> mountInfoLock(gMountInfoMtx);
```
Considering using a `lock_guard` to avoid explicit unlock of the mutex.

rolysr replied:
Done.

ralcolea commented on src/mount/mastercomm.cc:
```diff
@@ -1086,6 +1086,26 @@ void* fs_nop_thread(void *arg) {
 				free(inodespacket);
 			}
+			if (masterversion >= saunafsVersion(4, 9, 0)) {
+				std::unique_lock<std::mutex> mountInfoLock(gMountInfoMtx);
```
`mountInfoLock` is held while constructing the `mountInfoPacket` and send it through the network.
This can lead to contention if other threads frequently access  `gMountInfoMtx`.
Consider constructing the packet outside the lock to minimize the critical section.
```cpp
  std::string mountInfoStr;
  {
        std::unique_lock<std::mutex> mountInfoLock(gMountInfoMtx);
 		mountInfoStr = gMountInfo.getMountInfoStr();
   }
  auto message = cltoma::updateMountInfo::build(mountInfoStr);
```

rolysr replied:
Done.

ralcolea commented on src/mount/mastercomm.cc:
```diff
@@ -1086,6 +1086,26 @@ void* fs_nop_thread(void *arg) {
+				auto message = cltoma::updateMountInfo::build(mountInfoStr);
+
+				uint32_t messageLength = message.size();
+				uint8_t *mountInfoPacket = (uint8_t *)malloc(messageLength);
```
The `malloc` call for `mountInfoPacket` returns a nullptr pointer if allocation is not successfully, leading to a null pointer exception in the `memcpy` function.

Consider using a more modern and safe alternative to allocate the buffer:
```cpp
 std::vector<uint8_t> mountInfoPacket(messageLength);
 std::copy(message.begin(), message.end(), mountInfoPacket.begin());

if (tcptowrite(fd, mountInfoPacket.data(), messageLength, 1000) !=
    (int32_t)messageLength) {
	disconnect = true;
	safs_pretty_syslog(LOG_WARNING, "Failed to send mount info to master");
} else {
	master_stats_add(MASTER_BYTESSENT, messageLength);
	master_stats_inc(MASTER_PACKETSSENT);
}
```

rolysr replied:
Done.

ralcolea commented on src/mount/mount_info.cc:
```diff
@@ -192,6 +193,7 @@ std::string get_username_by_uid(uid_t uid) {
 #endif
 
 void set_all_mountpoint_arguments(int argc, char **argv) {
+	std::unique_lock<std::mutex> lock(gMountInfoMtx);
```
To reduce contention, it's a good practice to keep the critical section as minimum as possible.
The lock can created exactly before accesing `gMountInfo` instance.

rolysr replied:
Done.

ralcolea commented on src/mount/mount_info.cc:
```diff
@@ -201,6 +203,7 @@ void set_all_mountpoint_arguments(int argc, char **argv) {
 }
 
 void set_current_local_time() {
+	std::unique_lock<std::mutex> lock(gMountInfoMtx);
```
Same comment as before.

rolysr replied:
Done.

ralcolea commented on src/protocol/SFSCommunication.h:
```diff
@@ -1243,6 +1243,10 @@ enum class SugidClearMode {
 #define SAU_CLTOMA_REGISTER_CONFIG (1000U + 486U)
 /// config:STDSTRING
 
+//0x01E6
```
`0x01E6` tag is the same as the previous one.
Maybe `0x01E7` should be used?

rolysr replied:
Done.

ralcolea commented on tests/tools/valgrind-helgrind.supp:
```diff
@@ -60,6 +60,20 @@
+   fs_close_session
+   Helgrind:Misc
+   ...
+   fun:_Z16fs_close_sessionv
```
Adapt your suppresions like in [(36905fa)](https://github.com/leil-io/saunafs/pull/367/commits/36905fa0a3d02b074c43055456fb6726c5fa8e53) to avoid they will be reported by clang compiler.

Basically you need to remove from the expression the prefix starting with `Z` that reports GCC, but not clang and use a wildcard (**\'*'**) to replace the Z and the digits in the name of the function.

rolysr replied:
Done.

ralcolea reviewed (approved):
Very nice job @rolysr :+1: :fire: :muscle: 
Please see some minor suggestions to improve the final version before merging the changes.

lgsilva3087 commented on src/mount/mount_info.cc:
```diff
@@ -102,6 +102,7 @@ void mount_info_init(
 #endif
 	const std::string &username, int pid, const std::string &version,
 	const std::string &commitId) {
+	std::lock_guard<std::mutex> lock(gMountInfoMtx);
```
std::lock_guard is able to deduce the type from the mutex, please consider everywhere:
`std::lock_guard lock(gMountInfoMtx);`

rolysr replied:
Done.

lgsilva3087 commented on src/mount/mastercomm.cc:
```diff
@@ -1086,6 +1086,28 @@ void* fs_nop_thread(void *arg) {
 				}
 				free(inodespacket);
 			}
+			if (masterversion >= saunafsVersion(4, 9, 0)) {
```
Looks to often to send the info on every loop. Please consider to sent it only if it changes.

rolysr replied:
Done.

lgsilva3087 commented on src/mount/mastercomm.cc:
```diff
@@ -1086,6 +1086,28 @@ void* fs_nop_thread(void *arg) {
 				}
 				free(inodespacket);
 			}
+			if (masterversion >= saunafsVersion(4, 9, 0)) {
```
The previous parts could set disconnect to true, in that case, the call to tcptowrite should also fail. Consider checking it in the condition.

rolysr replied:
Done.

lgsilva3087 commented on src/mount/mount_info.cc:
```diff
@@ -40,7 +40,7 @@ std::string MountInfo::getVersion() const { return version_; }
 std::string MountInfo::getArguments() const { return arguments_; }
 std::map<std::string, std::string> MountInfo::getMountOptions() const { return *mountOptions_; }
-const std::string &MountInfo::getMountInfoStr() {
+const std::string MountInfo::getMountInfoStr() {
```
The const is not doing too much here with a string. Consider to remove it.

rolysr replied:
Done.

lgsilva3087 commented on src/mount/mount_info.cc:
```diff
@@ -40,7 +40,7 @@ std::string MountInfo::getVersion() const { return version_; }
 std::map<std::string, std::string> MountInfo::getMountOptions() const { return *mountOptions_; }
-const std::string &MountInfo::getMountInfoStr() {
+const std::string MountInfo::getMountInfoStr() {
 	if (mountInfoStr_.empty()) { buildMountInfoStr(); }
```
Is it built only once?

rolysr replied:
From the `getMountInfoStr`, yes, this ensures that the requested mount info is never empty. Anyway, it is also rebuild in other parts of the code like the `special open` and the `fs_nop_thread`.

lgsilva3087 reviewed (commented):
Please see my comments.

dmga44 commented on src/master/matoclserv.cc:
```diff
@@ -1257,18 +1258,29 @@ void matoclserv_chunks_health(matoclserventry *eptr, const uint8_t *data, uint32
+	uint32_t size,ileng,mount_info_leng,pleng,i;
 	uint8_t vmode;
+	uint8_t includeMountInfo = false;
+	uint8_t mount_info_leng_size = 4;
```
A small suggestion here, this line could be subtituted by something like 
`constexpr uint8_t mount_info_leng_size = sizeof(mount_info_leng);`
please consider it.

dmga44 replied:
As an extra suggestion, please check the common size of the `mountinfo` it could be more than enough to have it in a 16 bit integer (up to 65535 bytes size). In the client side you could consider this type change when necessary,

rolysr replied:
Considered, but at the end the approach was changed.

rolysr replied:
Considered but changed final approach.

dmga44 commented on src/master/matoclserv.cc:
```diff
@@ -1257,18 +1258,29 @@ void matoclserv_chunks_health(matoclserventry *eptr, const uint8_t *data, uint32
+	auto isMountInfoRequired = [](bool includeMountInfo, const matoclserventry *eaptr) -> bool {
+		return includeMountInfo && eaptr->version >= saunafsVersion(4, 9, 0) && eaptr->sesdata &&
+		       eaptr->sesdata->mountinfo.size() > 0;
+	};
```
I guess these lines over here are ok, but IMO the check `&& eaptr->version >= saunafsVersion(4, 9, 0)` is not necessary because only clients with version higher than 4.9 will have the `mountinfo` member of the session data populated.. Furthermore, the current v4.9.0 clients won't have it. Please consider it, you could also replace it after next release to something like v4.9.1 that will have it.

rolysr replied:
Done.

dmga44 commented on src/master/matoclserv.cc:
```diff
@@ -1905,6 +1927,14 @@ void matoclserv_register_config(matoclserventry *eptr, const uint8_t *data,
 }
 
+void matoclserv_update_mount_info(matoclserventry *eptr, const uint8_t *data, uint32_t length) {
+	std::string mount_info;
+	cltoma::updateMountInfo::deserialize(data, length, mount_info);
+	if (eptr->sesdata && eptr->sesdata->mountinfo != mount_info) {
+		eptr->sesdata->mountinfo = mount_info;
+	}
```
Consider to directly deserialize to the `mountinfo` member of the session data, after having checked that the `sesdata` exists.

rolysr replied:
Considered.

dmga44 commented on src/mount/mastercomm.cc:
```diff
@@ -1079,13 +1084,38 @@ void* fs_nop_thread(void *arg) {
 				free(inodespacket);
 			}
+
+			if (masterversion >= saunafsVersion(4, 9, 0) && !disconnect && gChangedTweaksValue) {
```
I think this part is also ok, just consider that for the current v4.9.0 master this package will be ignored, so clients having this feature will send this message to master (v.4.9.0) that won't understand it.

rolysr replied:
Done.

dmga44 reviewed (approved):
LGTM, but please consider my suggestions.

rolysr committed:
feat(admin): Add mount info list command

Add a new list-mount-info command to the saunafs-admin tool. It lists
every client’s mountpoint information for sessions open on a given
master server, making it easy to inspect active client connections.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

rolysr committed:
test: Add test for list-mount-info

Add a new test that verifies the list-mount-info command in the
saunafs-admin tool works as expected.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

dmga44 commented on src/cgi/sfs.cgi.in:
```diff
@@ -2327,10 +2348,11 @@ if "MS" in sectionset:
             else:
                 statscnt = struct.unpack(">H", data[0:2])[0]
-                pos = 2
+                pos = 2  
```
Please remove the couple of trailing spaces.

rolysr replied:
Done.

dmga44 reviewed (approved):
Great job @rolysr , please check my comment.

lgsilva3087 reviewed (approved):
LGTM
